File: plugins/rook-maintainer/scripts/sweep_prefetch.py
#!/usr/bin/env python3
"""Phase-0 metadata snapshot for a rook-triage or rook-code-review sweep.

Both call it — rook-triage's pipeline phase 0 and rook-code-review's
sweep phase 0 — so a change here must keep both callers working.

One batched GraphQL pass per corpus, written to <sweep-dir>/snapshot.json.
Every triager and reviewer agent and every dashboard generator consumes
this snapshot instead of fetching per-item metadata themselves — one
fetch, one consistent point-in-time view, ~100 fewer per-agent gh calls
per sweep.

PR items carry `files` (changed paths), which is what lets a sweep
orchestrator route reference files per PR without a per-PR fetch;
`gh pr list` cannot return them at any --json setting.

PR items carry a summarized statusCheckRollup ({passing, failing,
pending, total, failed[]}) — the ONLY source dashboards may use for CI
cells (deterministic passing/total; never parsed from agent prose).

Modes:
  snapshot      enumerate all OPEN items of --kind, or exactly --numbers
                (numbers mode also fetches closed/merged items, for
                regenerating dashboards of past sweeps)
  classify-refs scan <sweep-dir>/batch-*.json xlink/dup numbers and write
                refs-types.json (Issue vs PullRequest via
                issueOrPullRequest — never guessed)

Usage:
  python3 sweep_prefetch.py snapshot SWEEP_DIR --kind prs|issues
          [--numbers 1,2,3 | --numbers-file F] [--repo rook/rook]
  python3 sweep_prefetch.py classify-refs SWEEP_DIR [--repo rook/rook]

Needs authenticated gh (run with the sandbox disabled).
"""
import argparse
import datetime
import glob
import json
import os
import subprocess

PR_FIELDS = """
number title state isDraft updatedAt createdAt
author { login } authorAssociation baseRefName mergeable reviewDecision
additions deletions changedFiles
labels(first: 20) { nodes { name } }
assignees(first: 10) { nodes { login } }
files(first: 100) { pageInfo { hasNextPage endCursor } nodes { path } }
latestReviews(first: 20) { nodes { author { login } state } }
reviewRequests(first: 20) { nodes { requestedReviewer {
  ... on User { login } ... on Team { name } } } }
commits(last: 1) { nodes { commit { statusCheckRollup {
  state contexts(first: 100) { pageInfo { hasNextPage endCursor } nodes { __CTX__ } } } } } }
"""

CTX_NODE = """
__typename
... on CheckRun { name status conclusion
  checkSuite { databaseId createdAt workflowRun { workflow { name } } } }
... on StatusContext { context state }
"""
PR_FIELDS = PR_FIELDS.replace("__CTX__", CTX_NODE)

ISSUE_FIELDS = """
number title state updatedAt createdAt
author { login }
labels(first: 20) { nodes { name } }
assignees(first: 10) { nodes { login } }
comments { totalCount }
"""

PASS_CONCLUSIONS = {"SUCCESS", "NEUTRAL", "SKIPPED"}
FAIL_CONCLUSIONS = {"FAILURE", "TIMED_OUT", "CANCELLED", "ACTION_REQUIRED", "STARTUP_FAILURE"}


def gql(query):
    proc = subprocess.run(["gh", "api", "graphql", "-f", f"query={query}"],
                          capture_output=True, text=True, timeout=180)
    if proc.returncode != 0:
        raise RuntimeError(f"gh api graphql rc={proc.returncode} stderr={proc.stderr[:3000]}")
    data = json.loads(proc.stdout)
    if data.get("errors"):
        raise RuntimeError(f"GraphQL errors: {json.dumps(data['errors'])[:2000]}")
    return data["data"]


def suite_key(c):
    suite = c.get("checkSuite") or {}
    wf = ((suite.get("workflowRun") or {}).get("workflow") or {}).get("name")
    return wf or f"suite:{suite.get('databaseId')}", suite.get("createdAt") or ""


def classify_ctx_nodes(state, nodes, truncated):
    """Summarize context nodes the way GitHub's merge box does. The
    rollup keeps every CheckRun from every check suite ever run on the
    commit — a CI re-run months later under a changed workflow/matrix
    leaves the old generation's (differently-named) runs in the list,
    silently doubling job counts. So: keep only the NEWEST check suite
    per workflow, then dedupe by check name (re-run attempts within a
    suite), keeping the last occurrence."""
    newest = {}
    for c in nodes:
        if c["__typename"] != "CheckRun":
            continue
        key, created = suite_key(c)
        if key not in newest or created > newest[key]:
            newest[key] = created
    latest = {}
    for c in nodes:
        if c["__typename"] == "CheckRun":
            key, created = suite_key(c)
            if created != newest.get(key):
                continue
            latest[("run", c.get("name") or "?")] = c
        else:
            latest[("ctx", c.get("context") or "?")] = c
    passing = failing = pending = 0
    failed = []
    for (kind, name), c in latest.items():
        if kind == "run":
            if c.get("status") != "COMPLETED":
                pending += 1
            elif c.get("conclusion") in PASS_CONCLUSIONS:
                passing += 1
            elif c.get("conclusion") in FAIL_CONCLUSIONS:
                failing += 1
                failed.append(name)
            else:
                pending += 1
        else:
            st = c.get("state")
            if st == "SUCCESS":
                passing += 1
            elif st in ("FAILURE", "ERROR"):
                failing += 1
                failed.append(name)
            else:
                pending += 1
    return {"state": state, "passing": passing, "failing": failing,
            "pending": pending, "total": passing + failing + pending,
            "failed": failed, "truncated": truncated}


def summarize_ci(node):
    commits = node.get("commits", {}).get("nodes") or []
    rollup = (commits[0].get("commit") or {}).get("statusCheckRollup") if commits else None
    if not rollup:
        return {"state": None, "passing": 0, "failing": 0, "pending": 0,
                "total": 0, "failed": [], "truncated": False}
    ctx = rollup.get("contexts") or {}
    return classify_ctx_nodes(rollup.get("state"), ctx.get("nodes") or [],
                              bool((ctx.get("pageInfo") or {}).get("hasNextPage")))


def shape_pr(n):
    return {
        "number": n["number"], "title": n["title"], "state": n["state"],
        "isDraft": n["isDraft"], "updatedAt": n["updatedAt"], "createdAt": n["createdAt"],
        "author": (n.get("author") or {}).get("login", ""),
        "authorAssociation": n.get("authorAssociation"),
        "baseRefName": n.get("baseRefName"), "mergeable": n.get("mergeable"),
        "reviewDecision": n.get("reviewDecision"),
        "additions": n.get("additions"), "deletions": n.get("deletions"),
        "changedFiles": n.get("changedFiles"),
        "labels": [l["name"] for l in n["labels"]["nodes"]],
        "assignees": [a["login"] for a in n["assignees"]["nodes"]],
        "files": [f["path"] for f in n["files"]["nodes"]],
        "files_truncated": n["files"]["pageInfo"]["hasNextPage"],
        "reviews": {
            "latest": [{"login": (r.get("author") or {}).get("login", ""), "state": r["state"]}
                       for r in n["latestReviews"]["nodes"] if r.get("author")],
            "requested": [rr["requestedReviewer"].get("login") or rr["requestedReviewer"].get("name")
                          for rr in n["reviewRequests"]["nodes"] if rr.get("requestedReviewer")],
        },
        "ci": summarize_ci(n),
    }


def shape_issue(n):
    return {
        "number": n["number"], "title": n["title"], "state": n["state"],
        "updatedAt": n["updatedAt"], "createdAt": n["createdAt"],
        "author": (n.get("author") or {}).get("login", ""),
        "labels": [l["name"] for l in n["labels"]["nodes"]],
        "assignees": [a["login"] for a in n["assignees"]["nodes"]],
        "comments_total": n["comments"]["totalCount"],
    }


def fetch_open(owner, name, kind):
    field = "pullRequests" if kind == "prs" else "issues"
    inner = PR_FIELDS if kind == "prs" else ISSUE_FIELDS
    page = 25 if kind == "prs" else 100
    nodes, cursor = [], None
    while True:
        after = f', after: "{cursor}"' if cursor else ""
        q = (f'query {{ repository(owner: "{owner}", name: "{name}") {{ '
             f'{field}(states: OPEN, first: {page}{after}, '
             f'orderBy: {{field: CREATED_AT, direction: ASC}}) {{ '
             f'pageInfo {{ hasNextPage endCursor }} nodes {{ {inner} }} }} }} }}')
        block = gql(q)["repository"][field]
        nodes += block["nodes"]
        if not block["pageInfo"]["hasNextPage"]:
            return nodes
        cursor = block["pageInfo"]["endCursor"]


def fetch_numbers(owner, name, kind, numbers):
    field = "pullRequest" if kind == "prs" else "issue"
    inner = PR_FIELDS if kind == "prs" else ISSUE_FIELDS
    chunk = 15 if kind == "prs" else 75
    nodes = []
    for i in range(0, len(numbers), chunk):
        aliases = " ".join(f"n{x}: {field}(number: {x}) {{ {inner} }}"
                           for x in numbers[i:i + chunk])
        repo = gql(f'query {{ repository(owner: "{owner}", name: "{name}") {{ {aliases} }} }}')["repository"]
        nodes += [v for v in repo.values() if v]
    return nodes


def paginate_all_contexts(owner, name, number):
    nodes, cursor = [], None
    while True:
        after = f', after: "{cursor}"' if cursor else ""
        q = (f'query {{ repository(owner: "{owner}", name: "{name}") {{ '
             f'pullRequest(number: {number}) {{ commits(last: 1) {{ nodes {{ commit {{ '
             f'statusCheckRollup {{ state contexts(first: 100{after}) {{ '
             f'pageInfo {{ hasNextPage endCursor }} nodes {{ {CTX_NODE} }} }} }} }} }} }} }} }} }}')
        rollup = gql(q)["repository"]["pullRequest"]["commits"]["nodes"][0]["commit"]["statusCheckRollup"]
        block = rollup["contexts"]
        nodes += block["nodes"]
        if not block["pageInfo"]["hasNextPage"]:
            return rollup.get("state"), nodes
        cursor = block["pageInfo"]["endCursor"]


def paginate_all_files(owner, name, number):
    paths, cursor = [], None
    while True:
        after = f', after: "{cursor}"' if cursor else ""
        q = (f'query {{ repository(owner: "{owner}", name: "{name}") {{ '
             f'pullRequest(number: {number}) {{ files(first: 100{after}) {{ '
             f'pageInfo {{ hasNextPage endCursor }} nodes {{ path }} }} }} }} }}')
        block = gql(q)["repository"]["pullRequest"]["files"]
        paths += [f["path"] for f in block["nodes"]]
        if not block["pageInfo"]["hasNextPage"]:
            return paths
        cursor = block["pageInfo"]["endCursor"]


def cmd_snapshot(args):
    owner, name = args.repo.split("/", 1)
    if args.numbers or args.numbers_file:
        if args.numbers:
            nums = [int(x) for x in args.numbers.split(",") if x.strip()]
        else:
            nums = [int(l) for l in open(args.numbers_file) if l.strip()]
        nodes = fetch_numbers(owner, name, args.kind, sorted(set(nums)))
    else:
        nodes = fetch_open(owner, name, args.kind)
    shape = shape_pr if args.kind == "prs" else shape_issue
    items = {str(n["number"]): shape(n) for n in nodes}
    if args.kind == "prs":
        for num, item in items.items():
            if item["ci"].get("truncated"):
                state, ctx_nodes = paginate_all_contexts(owner, name, int(num))
                item["ci"] = classify_ctx_nodes(state, ctx_nodes, False)
            if item.get("files_truncated"):
                item["files"] = paginate_all_files(owner, name, int(num))
                item["files_truncated"] = False
    out = {
        "fetched_at": datetime.datetime.now(datetime.timezone.utc).isoformat(),
        "repo": args.repo, "kind": args.kind, "items": items,
    }
    os.makedirs(args.sweep_dir, exist_ok=True)
    path = os.path.join(args.sweep_dir, "snapshot.json")
    json.dump(out, open(path, "w"), indent=1)
    trunc = [k for k, v in items.items()
             if v.get("files_truncated") or (v.get("ci") or {}).get("truncated")]
    print(f"{len(items)} {args.kind} -> {path}" + (f" (truncated fields on: {trunc})" if trunc else ""))


def cmd_classify_refs(args):
    owner, name = args.repo.split("/", 1)
    nums = set()
    for b in sorted(glob.glob(os.path.join(args.sweep_dir, "batch-*.json"))):
        for it in json.load(open(b)):
            for x in (it.get("xlinks") or []) + (it.get("dups") or []):
                if isinstance(x, dict) and "number" in x:
                    nums.add(int(x["number"]))
    path = os.path.join(args.sweep_dir, "refs-types.json")
    types = json.load(open(path)) if os.path.exists(path) else {}
    todo = sorted(n for n in nums if str(n) not in types)
    for i in range(0, len(todo), 75):
        aliases = " ".join(f"n{x}: issueOrPullRequest(number: {x}) {{ __typename }}"
                           for x in todo[i:i + 75])
        repo = gql(f'query {{ repository(owner: "{owner}", name: "{name}") {{ {aliases} }} }}')["repository"]
        for k, v in repo.items():
            if v:
                types[k[1:]] = v["__typename"]
    json.dump(types, open(path, "w"), indent=1)
    print(f"{len(nums)} refs, {len(todo)} newly classified -> {path}")


def main():
    ap = argparse.ArgumentParser()
    sub = ap.add_subparsers(dest="cmd", required=True)
    s = sub.add_parser("snapshot")
    s.add_argument("sweep_dir")
    s.add_argument("--kind", choices=["prs", "issues"], required=True)
    s.add_argument("--numbers")
    s.add_argument("--numbers-file")
    s.add_argument("--repo", default="rook/rook")
    s.set_defaults(func=cmd_snapshot)
    c = sub.add_parser("classify-refs")
    c.add_argument("sweep_dir")
    c.add_argument("--repo", default="rook/rook")
    c.set_defaults(func=cmd_classify_refs)
    args = ap.parse_args()
    args.func(args)


if __name__ == "__main__":
    main()

File: plugins/rook-maintainer/tools/internal/sweepprefetch/sweepprefetch.go
// Package sweepprefetch takes the phase-0 metadata snapshot a rook-triage or
// rook-code-review sweep runs on: one batched GraphQL pass per corpus, written
// to <sweep-dir>/snapshot.json.
//
// Every triager, reviewer and dashboard generator reads that snapshot instead
// of fetching per-item metadata itself — one fetch, one consistent
// point-in-time view, ~100 fewer per-agent gh calls per sweep. Two payloads
// only this pass can supply:
//
//   - PR files (changed paths), which is what lets a sweep orchestrator route
//     reference files per PR without a per-PR fetch; `gh pr list` cannot
//     return them at any --json setting.
//   - a summarized statusCheckRollup, the ONLY source dashboards may use for
//     CI cells (deterministic passing/total, never parsed from agent prose).
//
// The written shape is a contract with consumers outside this module, so
// field names, nesting and null-vs-absent are all load-bearing.
package sweepprefetch

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/jhoblitt/rook-claude/plugins/rook-maintainer/tools/internal/ghx"
)

const ctxNode = `
__typename
... on CheckRun { name status conclusion
  checkSuite { databaseId createdAt workflowRun { workflow { name } } } }
... on StatusContext { context state }
`

const prFields = `
number title state isDraft updatedAt createdAt
author { login } authorAssociation baseRefName mergeable reviewDecision
additions deletions changedFiles
labels(first: 20) { nodes { name } }
assignees(first: 10) { nodes { login } }
files(first: 100) { pageInfo { hasNextPage endCursor } nodes { path } }
latestReviews(first: 20) { nodes { author { login } state } }
reviewRequests(first: 20) { nodes { requestedReviewer {
  ... on User { login } ... on Team { name } } } }
commits(last: 1) { nodes { commit { statusCheckRollup {
  state contexts(first: 100) { pageInfo { hasNextPage endCursor } nodes { ` + ctxNode + ` } } } } } }
`

const issueFields = `
number title state updatedAt createdAt
author { login }
labels(first: 20) { nodes { name } }
assignees(first: 10) { nodes { login } }
comments { totalCount }
`

// GraphQLFunc runs one query and unmarshals its data object into out.
type GraphQLFunc func(ctx context.Context, query string, out any) error

type Client struct {
	Repo  string
	Owner string
	Name  string

	// GraphQL and Now are injection points for tests; zero values use gh and
	// the wall clock.
	GraphQL GraphQLFunc
	Now     func() time.Time
}

func NewClient(repo string) (*Client, error) {
	owner, name, ok := strings.Cut(repo, "/")
	if !ok || owner == "" || name == "" {
		return nil, fmt.Errorf("repo %q is not owner/name", repo)
	}
	return &Client{Repo: repo, Owner: owner, Name: name}, nil
}

func (c *Client) query(ctx context.Context, q string, out any) error {
	if c.GraphQL != nil {
		return c.GraphQL(ctx, q, out)
	}
	return ghx.GraphQL(ctx, q, out)
}

func (c *Client) now() time.Time {
	if c.Now != nil {
		return c.Now()
	}
	return time.Now()
}

type SnapshotOptions struct {
	SweepDir string
	// Kind is "prs" or "issues".
	Kind string
	// ByNumber fetches exactly Numbers — including closed and merged items,
	// which is how a dashboard for a past sweep gets regenerated — instead of
	// enumerating every open item.
	ByNumber bool
	Numbers  []int
}

type SnapshotResult struct {
	Path      string
	Kind      string
	Count     int
	Truncated []string
}

func (r SnapshotResult) String() string {
	s := fmt.Sprintf("%d %s -> %s", r.Count, r.Kind, r.Path)
	if len(r.Truncated) == 0 {
		return s
	}
	quoted := make([]string, len(r.Truncated))
	for i, n := range r.Truncated {
		quoted[i] = "'" + n + "'"
	}
	return s + fmt.Sprintf(" (truncated fields on: [%s])", strings.Join(quoted, ", "))
}

type snapshotFile struct {
	FetchedAt string  `json:"fetched_at"`
	Repo      string  `json:"repo"`
	Kind      string  `json:"kind"`
	Items     *Object `json:"items"`
}

func (c *Client) Snapshot(ctx context.Context, opts SnapshotOptions) (SnapshotResult, error) {
	res := SnapshotResult{
		Path: filepath.Join(opts.SweepDir, "snapshot.json"),
		Kind: opts.Kind,
	}
	items := NewObject()
	if opts.Kind == "prs" {
		prs, err := c.snapshotPRs(ctx, opts)
		if err != nil {
			return res, err
		}
		for _, item := range prs {
			items.Set(strconv.Itoa(item.Number), item)
			if item.FilesTruncated || item.CI.Truncated {
				res.Truncated = append(res.Truncated, strconv.Itoa(item.Number))
			}
		}
	} else {
		nodes, err := c.fetchIssueNodes(ctx, opts)
		if err != nil {
			return res, err
		}
		for _, n := range nodes {
			item := shapeIssue(n)
			items.Set(strconv.Itoa(item.Number), item)
		}
	}

	out := snapshotFile{
		FetchedAt: c.now().UTC().Format("2006-01-02T15:04:05.000000-07:00"),
		Repo:      c.Repo,
		Kind:      opts.Kind,
		Items:     items,
	}
	data, err := Encode(out)
	if err != nil {
		return res, err
	}
	if err := os.MkdirAll(opts.SweepDir, 0o755); err != nil {
		return res, err
	}
	if err := os.WriteFile(res.Path, data, 0o644); err != nil {
		return res, err
	}
	res.Count = items.Len()
	return res, nil
}

func (c *Client) snapshotPRs(ctx context.Context, opts SnapshotOptions) ([]*PRItem, error) {
	var nodes []prNode
	var err error
	if opts.ByNumber {
		nodes, err = fetchNumbers[prNode](ctx, c, "pullRequest", prFields, 15, opts.Numbers)
	} else {
		nodes, err = fetchOpen[prNode](ctx, c, "pullRequests", prFields, 25)
	}
	if err != nil {
		return nil, err
	}
	items := dedupePRs(nodes)
	// The batched query caps contexts and files at one page each; anything
	// that hit the cap is re-fetched on its own so the counts stay exact.
	for _, item := range items {
		if item.CI.Truncated {
			state, ctxNodes, err := c.paginateContexts(ctx, item.Number)
			if err != nil {
				return nil, err
			}
			item.CI = ClassifyContexts(state, ctxNodes, false)
		}
		if item.FilesTruncated {
			files, err := c.paginateFiles(ctx, item.Number)
			if err != nil {
				return nil, err
			}
			item.Files = files
			item.FilesTruncated = false
		}
	}
	return items, nil
}

// dedupePRs collapses repeated PR numbers exactly as the items map they end up
// in would: the last node wins, at the position the number was first seen.
//
// A cursor walk can hand back the same PR on two pages when the corpus shifts
// under it. Deduping has to precede the per-item re-fetch below: run that
// re-fetch per node instead and the duplicate burns a second gh call, and the
// copy that survives carries CI state and files from that later fetch rather
// than from the single pass every other item gets.
func dedupePRs(nodes []prNode) []*PRItem {
	items := make([]*PRItem, 0, len(nodes))
	at := make(map[int]int, len(nodes))
	for _, n := range nodes {
		item := shapePR(n)
		if i, seen := at[item.Number]; seen {
			items[i] = item
			continue
		}
		at[item.Number] = len(items)
		items = append(items, item)
	}
	return items
}

func (c *Client) fetchIssueNodes(ctx context.Context, opts SnapshotOptions) ([]issueNode, error) {
	if opts.ByNumber {
		return fetchNumbers[issueNode](ctx, c, "issue", issueFields, 75, opts.Numbers)
	}
	return fetchOpen[issueNode](ctx, c, "issues", issueFields, 100)
}

type repoWrapper struct {
	Repository map[string]json.RawMessage `json:"repository"`
}

type nodeBlock[T any] struct {
	PageInfo pageInfo `json:"pageInfo"`
	Nodes    []T      `json:"nodes"`
}

func fetchOpen[T any](ctx context.Context, c *Client, field, inner string, page int) ([]T, error) {
	var nodes []T
	cursor := ""
	for {
		q := fmt.Sprintf("query { repository(owner: %s, name: %s) { "+
			"%s(states: OPEN, first: %d%s, orderBy: {field: CREATED_AT, direction: ASC}) { "+
			"pageInfo { hasNextPage endCursor } nodes { %s } } } }",
			quote(c.Owner), quote(c.Name), field, page, after(cursor), inner)
		raw, err := c.repoField(ctx, q, field)
		if err != nil {
			return nil, err
		}
		var block nodeBlock[T]
		if err := json.Unmarshal(raw, &block); err != nil {
			return nil, fmt.Errorf("decoding %s page: %w", field, err)
		}
		nodes = append(nodes, block.Nodes...)
		if !block.PageInfo.HasNextPage {
			return nodes, nil
		}
		cursor = block.PageInfo.EndCursor
	}
}

func fetchNumbers[T any](ctx context.Context, c *Client, field, inner string, chunk int, numbers []int) ([]T, error) {
	var nodes []T
	for start := 0; start < len(numbers); start += chunk {
		batch := numbers[start:min(start+chunk, len(numbers))]
		var aliases strings.Builder
		for i, n := range batch {
			if i > 0 {
				aliases.WriteString(" ")
			}
			fmt.Fprintf(&aliases, "n%d: %s(number: %d) { %s }", n, field, n, inner)
		}
		q := fmt.Sprintf("query { repository(owner: %s, name: %s) { %s } }",
			quote(c.Owner), quote(c.Name), aliases.String())
		var wrap repoWrapper
		if err := c.query(ctx, q, &wrap); err != nil {
			return nil, err
		}
		for _, n := range batch {
			raw, ok := wrap.Repository[fmt.Sprintf("n%d", n)]
			if !ok || isNull(raw) {
				continue
			}
			var node T
			if err := json.Unmarshal(raw, &node); err != nil {
				return nil, fmt.Errorf("decoding %s %d: %w", field, n, err)
			}
			nodes = append(nodes, node)
		}
	}
	return nodes, nil
}

func (c *Client) paginateContexts(ctx context.Context, number int) (*string, []ContextNode, error) {
	var nodes []ContextNode
	cursor := ""
	for {
		q := fmt.Sprintf("query { repository(owner: %s, name: %s) { "+
			"pullRequest(number: %d) { commits(last: 1) { nodes { commit { "+
			"statusCheckRollup { state contexts(first: 100%s) { "+
			"pageInfo { hasNextPage endCursor } nodes { %s } } } } } } } } }",
			quote(c.Owner), quote(c.Name), number, after(cursor), ctxNode)
		var wrap struct {
			Repository *struct {
				PullRequest *struct {
					Commits struct {
						Nodes []struct {
							Commit struct {
								StatusCheckRollup *struct {
									State    *string                `json:"state"`
									Contexts nodeBlock[ContextNode] `json:"contexts"`
								} `json:"statusCheckRollup"`
							} `json:"commit"`
						} `json:"nodes"`
					} `json:"commits"`
				} `json:"pullRequest"`
			} `json:"repository"`
		}
		if err := c.query(ctx, q, &wrap); err != nil {
			return nil, nil, err
		}
		if wrap.Repository == nil || wrap.Repository.PullRequest == nil {
			return nil, nil, fmt.Errorf("graphql response for pull request %d contexts carried no repository.pullRequest", number)
		}
		commits := wrap.Repository.PullRequest.Commits.Nodes
		if len(commits) == 0 || commits[0].Commit.StatusCheckRollup == nil {
			return nil, nil, fmt.Errorf("pull request %d has no status check rollup to paginate", number)
		}
		rollup := commits[0].Commit.StatusCheckRollup
		nodes = append(nodes, rollup.Contexts.Nodes...)
		if !rollup.Contexts.PageInfo.HasNextPage {
			return rollup.State, nodes, nil
		}
		cursor = rollup.Contexts.PageInfo.EndCursor
	}
}

func (c *Client) paginateFiles(ctx context.Context, number int) ([]string, error) {
	paths := []string{}
	cursor := ""
	for {
		q := fmt.Sprintf("query { repository(owner: %s, name: %s) { "+
			"pullRequest(number: %d) { files(first: 100%s) { "+
			"pageInfo { hasNextPage endCursor } nodes { path } } } } }",
			quote(c.Owner), quote(c.Name), number, after(cursor))
		var wrap struct {
			Repository *struct {
				PullRequest *struct {
					Files nodeBlock[struct {
						Path string `json:"path"`
					}] `json:"files"`
				} `json:"pullRequest"`
			} `json:"repository"`
		}
		if err := c.query(ctx, q, &wrap); err != nil {
			return nil, err
		}
		// Without this the null unmarshals to an empty page that reads as
		// "pagination finished", and the caller replaces a real file list with
		// an empty one flagged complete.
		if wrap.Repository == nil || wrap.Repository.PullRequest == nil {
			return nil, fmt.Errorf("graphql response for pull request %d files carried no repository.pullRequest", number)
		}
		block := wrap.Repository.PullRequest.Files
		for _, f := range block.Nodes {
			paths = append(paths, f.Path)
		}
		if !block.PageInfo.HasNextPage {
			return paths, nil
		}
		cursor = block.PageInfo.EndCursor
	}
}

func (c *Client) repoField(ctx context.Context, q, field string) (json.RawMessage, error) {
	var wrap repoWrapper
	if err := c.query(ctx, q, &wrap); err != nil {
		return nil, err
	}
	raw, ok := wrap.Repository[field]
	if !ok || isNull(raw) {
		return nil, fmt.Errorf("graphql response carried no repository.%s", field)
	}
	return raw, nil
}

func after(cursor string) string {
	if cursor == "" {
		return ""
	}
	return ", after: " + quote(cursor)
}

// quote renders a GraphQL string literal; its escaping rules are JSON's.
func quote(s string) string {
	b, err := json.Marshal(s)
	if err != nil {
		return `""`
	}
	return string(b)
}

func isNull(raw json.RawMessage) bool {
	return string(bytes.TrimSpace(raw)) == "null"
}
